### ISSUE TITLE
Mail Rules improvements

### ISSUE THREAD

**Issue by danielbayley** [nylas/N1#1074]
* [ ] Nested rules (or some way to combine them for example ‘any of these conditions but none of these’
* [ ] More rule criteria
* [ ] Ability to rearrange rules with drag & drop
* [ ] Ability to quickly duplicate a rule
* [ ] Toggle dis/enable button for each rule
* [ ] Test run (maybe a ’Test’ button for each rule)
* [ ] Ability to group rules together

**Comment by mbilker** [comment]
I agree with this issue. Could you state some rule criteria you want? Test run like how Gmail allows you to see what messages would be impacted by a rule?

**Comment by danielbayley** [comment]
> Could you state some rule criteria you want

Well the main ones I can think of at the moment would be…

- is/not in folder
- is/not in rule
- is older/newer than
- is/not in contacts group (see #406)
- is/isn’t starred

Also the basic ability to combine rules would be a major one, like ‘all of these conditions **and** any of these… any of these, but **none** of these’ etc…

**Comment by bengotow** [comment]
Hmm, sounds like a good list—I think 1, 3, and 5 above should be straightforward to add. I'm hesitant to allow rules to be dependent on the execution / conditions of other rules because I think average users will have a hard time following dependency graphs of rules, and it adds the possibility of cycles in our rule interpreter, etc. That said, we might be able to add an "Advanced Mode" to that view :-)

**Comment by mbilker** [comment]
Is this a feature to implement on the client side or the Sync Engine side? There may be race conditions if multiple instances of N1 use the same rules.

**Comment by danielbayley** [comment]
What file/folder are the mail rules actually stored in by the way?

**Comment by mbilker** [comment]
@danielbayley They are stored inside `edgehill.db` as `JSONBlob`s.

**Comment by danielbayley** [comment]
Wouldn’t they be better as a separate `.json` file?

**Comment by bengotow** [comment]
That'd be fine also—they're in the database because 1) they're not really human-editable since they reference folder / label IDs, and 2) they are automatically synced between windows if they're in the database. We may actually move them to the cloud metadata store that's being built into the sync engine so they're available / editable across all your machines using N1, and could eventually be applied server side.

**Comment by jackiehluo** [comment]
Another request: https://nylas.zendesk.com/agent/tickets/4215

>The main thing would be a list of 'messages that matched this rule' straight below the rule. Gmail & inbox do this (although their rules work really badly).
>something showing if the regex isn't correct (not sure which engine you're using - the JS one?) >being allowed to put a /i (case insensitive flag) on the regex would be great as well.

**Comment by brunohq** [comment]
Here's another one, and for me the most needed one: https://github.com/nylas/N1/issues/2229

>  Mail Rule for Forwarding - It would be super helpful if I could set up a mail rule to forward to another address.
